### PR TITLE
feat: enhance JSON file output for experimental IaC [CC-679]

### PR DESCRIPTION
**Pull request by teodora-sandu**
- [x] Follows [CONTRIBUTING](https://github.com/snyk/snyk/blob/master/.github/CONTRIBUTING.md) rules

#### What does this PR do?
This PR enhances the JSON file output for the local execution (experimental) flow of IaC so it matches better with the original flow. 
[CC-679]

#### Where should the reviewer start?
The `formatScanResult` function in `src/cli/commands/test/iac-local-execution/results-formatter.ts` creates the structure and contents that will end up in the JSON file output. The changes in `src/cli/commands/test/iac-local-execution/types.ts` are the corresponding type changes for the new fields.

#### How should this be manually tested?
1. Download the `deployment_singleissue.yaml` attached to the JIRA task under the cloned `snyk` repo
2. Build the CLI using `npm run build`
3. Get the original JSON file output: `node ./dist/cli iac test deployment_singleissue.yaml --json-file-output=original.json`
4. Get the experimental JSON file output: `node ./dist/cli iac test deployment_singleissue.yaml --experimental --json-file-output=experimental.json`
5. Compare the two resulting files
6. Run from another folder to make sure the `targetFilePath` is correct: `cd src; node ../dist/cli iac test ../deployment_singleissue.yaml --experimental --json-file-output=experimental.json`

#### Any background context you want to provide?
The original flow uses default values for some of the fields and computes the `projectName` and `targetFilePath` from the scanned file path. Some fields relating to the org are still missing and will be covered by https://snyksec.atlassian.net/browse/CC-772

#### What are the relevant tickets?
[CC-679](https://snyksec.atlassian.net/browse/CC-679): https://snyksec.atlassian.net/browse/CC-679

#### Screenshots
Screenshot 2021-04-13 at 13 38 08
Screenshot 2021-04-13 at 13 38 26

[CC-679]: https://snyksec.atlassian.net/browse/CC-679

### CODE REVIEW

File: src/cli/commands/test/iac-local-execution/results-formatter.ts [hunk 1]
```diff
@@ -90,16 +94,25 @@ function formatScanResult(
...
     },
     isPrivate: true,
     packageManager: engineTypeToProjectType[scanResult.engineType],
     targetFile: scanResult.filePath,
+    targetFilePath,
+    vulnerabilities: [],
+    dependencyCount: 0,
+    licensesPolicy: null,
+    ignoreSettings: null,
+    projectName: path.basename(path.dirname(targetFilePath)),
```
Comment (teodora-sandu) [inline src/cli/commands/test/iac-local-execution/results-formatter.ts:115]: As per: https://github.com/snyk/snyk/blob/bf8f901b6bd53161d98173f958483b421d7e9ed7/src/lib/snyk-test/run-iac-test.ts#L107

File: src/cli/commands/test/iac-local-execution/results-formatter.ts [hunk 2]
```diff
@@ -90,16 +94,25 @@ function formatScanResult(
...
         formattedIssues,
         severityThreshold,
       ),
+      projectType: projectTypeByFileType[scanResult.fileType],
     },
     isPrivate: true,
     packageManager: engineTypeToProjectType[scanResult.engineType],
     targetFile: scanResult.filePath,
+    targetFilePath,
+    vulnerabilities: [],
```
Comment (teodora-sandu) [inline src/cli/commands/test/iac-local-execution/results-formatter.ts:111]: As per: https://github.com/snyk/snyk/blob/bf8f901b6bd53161d98173f958483b421d7e9ed7/src/lib/snyk-test/run-iac-test.ts#L32-L35

File: src/cli/commands/test/iac-local-execution/results-formatter.ts [hunk 3]
```diff
@@ -90,16 +94,25 @@ function formatScanResult(
       lineNumber,
     };
   });
+
+  const targetFilePath = path.resolve(scanResult.filePath, '.');
```
Comment (teodora-sandu) [inline src/cli/commands/test/iac-local-execution/results-formatter.ts:98]: The original CLI flow computes this slightly differently but from testing it I can see that the path it renders is incorrect:
https://github.com/snyk/snyk/blob/c1814531406981466a411ec3b82ba7b8c53ffdd6/src/lib/snyk-test/run-iac-test.ts#L62-L64

```
  "targetFilePath": "/Users/teodorasandu/Documents/repos/snyk/deployment_singleissue.yaml/Users/teodorasandu/Documents/repos/snyk/deployment_singleissue.yaml",
```

Comment (aron) [reply src/cli/commands/test/iac-local-execution/results-formatter.ts:98]: This is a long standing bug, the structure of `options.path` varies depending on how the paths are provided to the CLI. So worth trying the following inputs:
```
snyk iac test --experimental # all files
snyk iac test --experimental ./ # providing a directory path
snyk iac test --experimental ./file1.tf ./file2.tf # providing individual files
```

**Comment by aron** [comment]
Is this intended as an incremental step? I see that a lot of fields are now present but a few are still missing when I compare the current flow with the `--experimental` flow: 
image

**Comment by teodora-sandu** [comment]
@aron Yes this is incremental and the fields relating to the org metadata will be added under https://snyksec.atlassian.net/browse/CC-772

**Review by aron — APPROVED** [review]
Played around with it and looks good. The `targetFile` and `targetFilePath` still have strange values depending on the path provided, for example if you provide an absolute path then both `targetFile` and `targetFilePath` have the same value but I think that's outside of the scope of this ticket.

File: src/cli/commands/test/iac-local-execution/results-formatter.ts (hunk 3 above)
Comment (teodora-sandu) [reply src/cli/commands/test/iac-local-execution/results-formatter.ts:98]: The results look fine to me:
- `snyk-dev iac test --experimental --json-file-output=experimental-all-local.json # all files`:
```
[
  {
    ...
    "targetFile": "/Users/teodorasandu/Documents/repos/cloud-native-configs/k8s-configs/multi-file.yaml",
    "targetFilePath": "/Users/teodorasandu/Documents/repos/cloud-native-configs/k8s-configs/multi-file.yaml",
    "projectName": "k8s-configs",
    "path": "/Users/teodorasandu/Documents/repos/cloud-native-configs",
    ...
  }
]
```
- `snyk-dev iac test --experimental ./  --json-file-output=experimental-dir-local.json # providing a directory path`:
```
[
  {
    ...
    "targetFile": "/Users/teodorasandu/Documents/repos/cloud-native-configs/k8s-configs/multi-file.yaml",
    "targetFilePath": "/Users/teodorasandu/Documents/repos/cloud-native-configs/k8s-configs/multi-file.yaml",
    "projectName": "k8s-configs",
    "path": "./",
    ...
  }
]
```
- `snyk-dev iac test --experimental k8s-configs/multi-file.yaml k8s-configs/privilegedPod.yaml --json-file-output=experimental-indiv-local.json # providing individual files`:
```
[
  {
    ...
    "targetFile": "k8s-configs/multi-file.yaml",
    "targetFilePath": "/Users/teodorasandu/Documents/repos/cloud-native-configs/k8s-configs/multi-file.yaml",
    "projectName": "k8s-configs",
    "path": "k8s-configs/multi-file.yaml",
     ...
  }
]
```

Comment (aron) [reply src/cli/commands/test/iac-local-execution/results-formatter.ts:98]: 👍 `targetFilePath` is good, the `targetFile` is sometimes a partial path and sometimes an absolute path, to be honest I'm not sure what the correct value should be, it feels like it should be just the filename...

Comment (aron) [reply src/cli/commands/test/iac-local-execution/results-formatter.ts:98]: To be clear, this isn't broken by your PR, just something that we need to fix in the near future.

**Pull request merged**